internal/verifier/check.go

```diff
@@ -81,44 +76,83 @@ func (verifier *Verifier) CheckWorker(ctx context.Context) error {
 
 	verifier.writeStringBuilder(genStartReport)
 
-	waitForTaskCreation := 0
+	cancelableCtx, canceler := context.WithCancelCause(ctxIn)
+	eg, ctx := errgroup.WithContext(cancelableCtx)
 
-	for {
+	// If the change stream fails, everything should stop.
+	eg.Go(func() error {
 		select {
 		case err := <-verifier.changeStreamErrChan:
-			cancel()
 			return errors.Wrap(err, "change stream failed")
 		case <-ctx.Done():
-			cancel()
 			return nil
-		default:
 		}
+	})
 
-		verificationStatus, err := verifier.GetVerificationStatus()
-		if err != nil {
-			verifier.logger.Error().Msgf("Failed getting verification status: %v", err)
-		}
+	// Start the worker threads.
+	for i := 0; i < verifier.numWorkers; i++ {
+		eg.Go(func() error {
+			return errors.Wrapf(
+				verifier.work(ctx, i),
+				"worker %d failed",
+				i,
+			)
+		})
+		time.Sleep(10 * time.Millisecond)
+	}
+
+	waitForTaskCreation := 0
 
-		if waitForTaskCreation%2 == 0 {
-			if generation > 0 || verifier.gen0PendingCollectionTasks.Load() == 0 {
-				verifier.PrintVerificationSummary(ctx, GenerationInProgress)
+	succeeded := false
+
+	eg.Go(func() error {
+		for {
+			verificationStatus, err := verifier.GetVerificationStatus(ctx)
+			if err != nil {
+				return errors.Wrapf(
+					err,
+					"failed to retrieve status of generation %d's tasks",
+					generation,
+				)
 			}
-		}
 
-		//wait for task to be created, if none of the tasks found.
-		if verificationStatus.AddedTasks > 0 || verificationStatus.ProcessingTasks > 0 || verificationStatus.RecheckTasks > 0 {
-			waitForTaskCreation++
-			time.Sleep(verifier.verificationStatusCheckInterval)
-		} else {
-			verifier.PrintVerificationSummary(ctx, GenerationComplete)
-			verifier.logger.Debug().Msg("Verification tasks complete")
-			cancel()
-			wg.Wait()
-			break
+			if waitForTaskCreation%2 == 0 {
+				if generation > 0 || verifier.gen0PendingCollectionTasks.Load() == 0 {
+					verifier.PrintVerificationSummary(ctx, GenerationInProgress)
+				}
+			}
+
+			// The generation continues as long as >=1 task for this generation is
+			// “added” or “pending”.
+			if verificationStatus.AddedTasks > 0 || verificationStatus.ProcessingTasks > 0 {
+				waitForTaskCreation++
+				time.Sleep(verifier.verificationStatusCheckInterval)
+			} else {
+				verifier.PrintVerificationSummary(ctx, GenerationComplete)
+				succeeded = true
+				canceler(errors.Errorf("generation %d succeeded", generation))
+				return nil
+			}
 		}
+	})
+
+	err := eg.Wait()
+
+	if succeeded {
```

Comment:
Why so?

I did try creating a sentry error and giving that as the Cause for the cancellation, but for some reason that sentry didn’t get preserved. The mongo-go driver sometimes fails to preserve cancellation causes; I figured this is just another example of that.

We could check errors.Is(context.Canceled), but that’s actually less fine-grained than the boolean because it would catch _any_ cancellation rather than just the one this function caused.

Reply:
Even better, thanks!